### PR TITLE
Incorporate Scrypto and Sparkz-util, update project to scala 2.13, and crosscompile for 2.12

### PR DESCRIPTION
**Pull request by i-skrypnyk**
https://horizenlabs.atlassian.net/browse/SDK-185

### CODE REVIEW

**Review by paologalligit — APPROVED** [review]
LGTM, great job!

File: examples/build.sbt [hunk 1]
```diff
@@ -5,18 +5,18 @@ libraryDependencies ++= Seq(
   "org.scalatest" %% "scalatest" % "3.2.12" % "test",
   "org.scalacheck" %% "scalacheck" % "1.16.0" % "test",
   "org.scalatestplus" %% "scalatestplus-scalacheck" % "3.1.0.0-RC2" % Test,
-  "org.scorexfoundation" %% "iodb" % "0.3.2",
+  //"org.scorexfoundation" %% "iodb" % "0.4.0",
```
Comment (i-Alex) [inline examples/build.sbt]: can be removed

File: scrypto/build.sbt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  "-source" :: "1.8" ::
+    "-target" :: "1.8" ::
+    Nil
+
+libraryDependencies ++= Seq(
+  "org.rudogma" %% "supertagged" % "1.5",
+  "com.google.guava" % "guava" % "23.0",
+  "com.typesafe.scala-logging" %% "scala-logging" % "3.9.2",
+  "org.whispersystems" % "curve25519-java" % "0.5.0",
+  "org.bouncycastle" % "bcprov-jdk15to18" % "1.66"
```
Comment (i-Alex) [inline scrypto/build.sbt]: let's use the most recent version and check later that SDK use the same one.
https://search.maven.org/artifact/org.bouncycastle/bcprov-jdk15on/1.70/jar
check on SDK side:
```
        <!-- NOTE: this comes from scrypto and contains a vulnerability, we use bcprov-jdk15on:1.69 below,
         this will be removed with merging scrypto into sparkz-core -->
        <exclusion>
          <artifactId>bcprov-jdk15to18</artifactId>
          <groupId>org.bouncycastle</groupId>
        </exclusion>
```

File: scrypto/build.sbt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+javacOptions ++=
+  "-source" :: "1.8" ::
+    "-target" :: "1.8" ::
+    Nil
+
+libraryDependencies ++= Seq(
+  "org.rudogma" %% "supertagged" % "1.5",
+  "com.google.guava" % "guava" % "23.0",
+  "com.typesafe.scala-logging" %% "scala-logging" % "3.9.2",
+  "org.whispersystems" % "curve25519-java" % "0.5.0",
```
Comment (i-Alex) [inline scrypto/build.sbt]: we will need to remove this dependency and reuse the bouncycastle one as we do in SDK.
Because of the licence that doesn't fit our requirements.
Will put it as a separate task into the backlog

File: sparkz-util/src/main/scala/sparkz/util/serialization/Reader.scala [hunk 4]
```diff
@@ -0,0 +1,140 @@
+package sparkz.util.serialization
+
+
+trait Reader {
```
Comment (i-Alex) [inline sparkz-util/src/main/scala/sparkz/util/serialization/Reader.scala:4]: good time to add `def getBoolean(): Boolean` method.
We had an issue before that we have `putBoolean` in the Writer, but miss the method in the Reader.
Update UTs as well

File: src/main/scala/sparkz/core/network/NetworkController.scala [hunk 5]
```diff
@@ -456,7 +456,7 @@ class NetworkController(settings: NetworkSettings,
           val myAddress = InetAddress.getAllByName(myHost)
 
           val listenAddresses = NetworkUtils.getListenAddresses(bindAddress)
-          val valid = listenAddresses.exists(myAddress.contains)
+          val valid = listenAddresses.exists(addr => myAddress.contains(addr.getAddress))
```
Comment (i-Alex) [inline src/main/scala/sparkz/core/network/NetworkController.scala:465]: was it a bug before?

File: src/main/scala/sparkz/core/transaction/account/PublicKeyNoncedBox.scala [hunk 6]
```diff
@@ -14,7 +14,7 @@ trait PublicKeyNoncedBox[PKP <: PublicKey25519Proposition] extends Box[PKP] {
   lazy val publicKey: PKP = proposition
 
   override def equals(obj: Any): Boolean = obj match {
-    case acc: PublicKeyNoncedBox[PKP] => java.util.Arrays.equals(acc.id, this.id) && acc.value == this.value
+    case acc: PublicKeyNoncedBox[_] => java.util.Arrays.equals(acc.id, this.id) && acc.value == this.value
```
Comment (i-Alex) [inline src/main/scala/sparkz/core/transaction/account/PublicKeyNoncedBox.scala:17]: why do we hide the type here? Which kind of error/warning has been appeared in 2.13?

**Review by i-Alex — CHANGES_REQUESTED** [review]
Good job!

File: src/main/scala/sparkz/core/network/NetworkController.scala (hunk 5 above)
Comment (i-skrypnyk) [reply src/main/scala/sparkz/core/network/NetworkController.scala:465]: Yes, and I'm pretty sure that this bug is caused by this line - https://horizenlabs.atlassian.net/browse/EVM-314

File: src/main/scala/sparkz/core/transaction/account/PublicKeyNoncedBox.scala (hunk 6 above)
Comment (i-skrypnyk) [reply src/main/scala/sparkz/core/transaction/account/PublicKeyNoncedBox.scala:17]: This and other similar places containing nested types are eliminated by erasure and are not checked. Somehow this warning was not treated as an error previously, but now it is

**Comment by i-Alex** [comment]
@i-skrypnyk pls, merge dev and resolve conflicts

File: scrypto/src/test/scala/sparkz/crypto/signing/SigningFunctionsSpecification.scala [hunk 7]
```diff
@@ -15,45 +15,23 @@ class SigningFunctionsSpecification extends AnyPropSpec
...
-        Curve25519.verify(sig, message2, keyPair._2) should not be true
+        Ed25519.verify(sig, message1, keyPair._2) shouldBe true
+        Ed25519.verify(sig, message1, keyPair2._2) should not be true
+        Ed25519.verify(sig, message2, keyPair._2) should not be true
 
       }
     }
   }
 
-  property("shared secret should be same for both parties ") {
```
Comment (paologalligit) [inline scrypto/src/test/scala/sparkz/crypto/signing/SigningFunctionsSpecification.scala]: Why doesn't the new `Ed25519` class have a `createSharedSecret` method? Is it something we want to drop?

Comment (i-skrypnyk) [reply scrypto/src/test/scala/sparkz/crypto/signing/SigningFunctionsSpecification.scala]: The old method Curve25519.createSharedSecret was not used anywhere, so I dropped it.

**Review by paologalligit — APPROVED** [review]

File: scrypto/src/main/scala/sparkz/crypto/signatures/Ed25519.scala [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+  override def createKeyPair(seed: Array[Byte]): (PrivateKey, PublicKey) = {
+    val privateKey: Ed25519PrivateKeyParameters = new Ed25519PrivateKeyParameters(Sha256.hash(seed), 0)
+    val publicKey: Ed25519PublicKeyParameters = privateKey.generatePublicKey
+
+    (PrivateKey @@ privateKey.getEncoded, PublicKey @@ publicKey.getEncoded)
+  }
+
+  override def sign(privateKey: PrivateKey, message: MessageToSign): Signature = {
+    val signature = new Array[Byte](64)
```
Comment (i-Alex) [inline scrypto/src/main/scala/sparkz/crypto/signatures/Ed25519.scala:22]:
```suggestion
    val signature = new Array[Byte](SignatureLength)
```

**Review by i-Alex — CHANGES_REQUESTED** [review]
Check `sparkz-examples` module, seems we have removed `iodb` dependency that is required to build the project

File: examples/src/test/scala/hybrid/state/SimpleBoxTransactionSpecification.scala [hunk 9]
```diff
@@ -31,16 +31,16 @@ class SimpleBoxTransactionSpecification extends AnyPropSpec
...
-        PublicKey25519NoncedBox(PublicKey25519Proposition(PublicKey @@ Base58.decode("GcVQWfSWUKoPuxbhoqx18hD4JKs2L1cvVvJZFzXKWaQ2").get), Nonce @@ 8815511758437892756L, Value @@ 100000L),
-        PublicKey25519NoncedBox(PublicKey25519Proposition(PublicKey @@ Base58.decode("99NTyZ796bpvwLLhMmsfwo8J3Wu3rUioUQsHE9CSYQKz").get), Nonce @@ -7463379991470698374L, Value @@ 100000L),
-        PublicKey25519NoncedBox(PublicKey25519Proposition(PublicKey @@ Base58.decode("3zFqfiRPEoshgaZY7qCcSk6mihDhgnGodBDgqP92stci").get), Nonce @@ -3238799622036337810L, Value @@ 100000L),
-        PublicKey25519NoncedBox(PublicKey25519Proposition(PublicKey @@ Base58.decode("2hw8D3T2Jrf7QZ9k53gDxDWjrnCXLLtDv1oonKGzKw74").get), Nonce @@ -6037688402013706435L, Value @@ 100000L),
-        PublicKey25519NoncedBox(PublicKey25519Proposition(PublicKey @@ Base58.decode("5zv52oTnDQas6WWnftRRhtZiudNaNTJ72WZWfDjRtKCQ").get), Nonce @@ -8390758926615640787L, Value @@ 100000L)
+        PublicKey25519NoncedBox(PublicKey25519Proposition(PublicKey @@ Base58.decode("8EjkXVSTxMFjCvNNsTo8RBMDEVQmk7gYkW4SCDuvdsBG").get), Nonce @@ -8612101699978236906L, Value @@ 100000L),
```
Comment (i-Alex) [inline examples/src/test/scala/hybrid/state/SimpleBoxTransactionSpecification.scala:34]: what is the reason of the change of the Nonce?

**Review by i-Alex — CHANGES_REQUESTED** [review]
I still have some issues with building the `examples.persistence` because of `ByteArrayWrapper` and `QuickStore` classes. They use some missing dependencies: ` com.sun.xml.internal.messaging.saaj.util.ByteOutputStream` and `javax.xml.bind.DatatypeConverter`. I guess we don't need them at all.

File: examples/src/test/scala/hybrid/state/SimpleBoxTransactionSpecification.scala (hunk 9 above)
Comment (i-Alex) [reply examples/src/test/scala/hybrid/state/SimpleBoxTransactionSpecification.scala:34]: because we changed the key generator in Sparkz: replaced Curve25519 class with SDK one (Ed25519)

**Review by i-Alex — APPROVED** [review]
